### PR TITLE
executor: Should not show the table attribute that the table has been dropped by user in information_schema.attributes

### LINKED ISSUE #29686: Should not show the table attribute that the table has been dropped by user in information_schema.attributes
## Enhancement
After drop one table, the attribute will wait for next gc time to remove it, so that if needed, user can recover the table.

Here is the testing logs,
2021-11-10T18:41:36.252+0800	INFO	oltppkg/util.go:23	Step 1: Prepare one table
2021-11-10T18:41:36.562+0800	INFO	oltppkg/workload.go:228	drop database if exists labeldb
2021-11-10T18:41:36.827+0800	INFO	oltppkg/workload.go:233	create database if exists labeldb
2021-11-10T18:41:37.222+0800	INFO	oltppkg/workload.go:228	drop database if exists labeldb
2021-11-10T18:41:37.834+0800	INFO	oltppkg/workload.go:233	create database if exists labeldb
2021-11-10T18:41:38.399+0800	INFO	pd-basic/attibue_test.go:47	Create table if not exists sbtest1

   Expected
        <[]string | len:2, cap:2>: [
            "schema/labeldb/sbtest1",
            "schema/labeldb/sbtest99",   << Created by last test.
        ]
    to have length 1

    /Users/jianghongmei/Documents/GitHub/endless_private_master/testcase/pd/pd-basic/attibue_test.go:136

### PR DESCRIPTION
**Pull request by hehechen**
Issue Number: close #29686

Problem Summary:
Should not show the table attribute that the table has been dropped by user in information_schema.attributes.

### What is changed and how it works?
In `setDataForAttributes`, get tableID in rules and check if the corresponding table exists.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
Test drop/recreate table, rename table, truncate partition, recover table
image
image
image

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by xhebox — APPROVED** [review]

File: executor/infoschema_reader.go [hunk 1]
```diff
@@ -2871,5 +2884,53 @@ func checkRule(rule *label.Rule) (dbName, tableName string, err error) {
 	}
 	dbName = s[1]
 	tableName = s[2]
+	if len(s) > 3 {
+		partitionName = s[3]
+	} else {
```
Comment (rleungx) [inline executor/infoschema_reader.go:2889]: no need else

File: executor/infoschema_reader.go [hunk 2]
```diff
@@ -2871,5 +2884,53 @@ func checkRule(rule *label.Rule) (dbName, tableName string, err error) {
...
+		partitionName = ""
+	}
+	return
+}
+
+func decodeTableIDFromRule(rule *label.Rule) (tableID int64, err error) {
+	data := rule.Data[0]
```
Comment (rleungx) [inline executor/infoschema_reader.go:2896]: prefer to add a check for the length.

File: executor/infoschema_reader.go [hunk 3]
```diff
@@ -2871,5 +2884,53 @@ func checkRule(rule *label.Rule) (dbName, tableName string, err error) {
...
+	dataMap, ok := data.(map[string]interface{})
+	if !ok {
+		err = errors.New("get the label rules failed")
+		return
+	}
+	key, err := hex.DecodeString(fmt.Sprintf("%s", dataMap["start_key"]))
+	if err != nil {
+		err = errors.New("decode key from start_key in rule failed")
```
Comment (rleungx) [inline executor/infoschema_reader.go:2904]: How about printing the key?

File: executor/infoschema_reader.go [hunk 4]
```diff
@@ -2871,5 +2884,53 @@ func checkRule(rule *label.Rule) (dbName, tableName string, err error) {
...
+		return
+	}
+	_, bs, err := codec.DecodeBytes(key, nil)
+	if err == nil {
+		key = bs
+	}
+	tableID = tablecodec.DecodeTableID(key)
+	if tableID == 0 {
+		err = errors.New("decode tableID from key failed")
```
Comment (rleungx) [inline executor/infoschema_reader.go:2913]: ditto

**Review by rleungx — APPROVED** [review]

**Pull request merged**